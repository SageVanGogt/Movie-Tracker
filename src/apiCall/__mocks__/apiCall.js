import mockData from './../../mockData/mockData';
const fetchRecentFilms = jest.fn().mockImplementation(() => Promise.resolve(mockData))
const addUserFetch = jest.fn().mockImplementation(() => Promise.resolve({
    name: "",
    email: "great",
    password: ""
}))

export {fetchRecentFilms, addUserFetch};

