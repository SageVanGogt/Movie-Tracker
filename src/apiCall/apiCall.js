import key from './apiKey';

const fetchRecentFilms = async () => {
  const url = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&primary_release_date.gte=2014-09-15&primary_release_date.lte=2014-10-22`;
  try{
    const response = await fetch(url);
    const filmData = await response.json();
    return filmData;
  } catch(err){
    const error = 'Failed to fetch data'
    throw error;
  }
}

const addUserFetch = async (user) => {
  const url = "http://localhost:3000/api/users/new"
  try {
    const response = await fetch(url, {
      method: "POST",
      body: JSON.stringify(user),
      headers: {
        Accept: 'application/json',
        "Content-Type": "application/json"}
    })
    const userData = await response.json();
    return userData;
  } catch(err){
    const error = 'Failed to fetch data'
    throw error;
  }
}

const fetchUser = async (user) => {
  const email = user.email.toLowerCase();
  const  userBody = {
    method: 'POST',
    body: JSON.stringify({ email, password: user.password }),
    headers: {"Content-Type": "application/json"}
  }
  const url = 'http://localhost:3000/api/users';
  try {
    const response = await fetch(url, userBody)
    const userData = await response.json();
    return userData;
  } catch (err) {
    const error = 'Failed to fetch data'
    throw error;
  }
}

const postFavoriteToDb = async (favoriteInfo, user) => {
  const url = "http://localhost:3000/api/users/favorites/new";
  const postData = {
    method: 'POST',
    body: JSON.stringify({user_id: user.user_id, ...favoriteInfo}),
    headers: {
      Accept: 'application/json',
      "Content-Type": "application/json"
    }
  }
  try {
    await fetch(url, postData);
  } catch(err) {
    const error = "Failed to post favorite";
    throw error;
  }
}

const getUserFavorites = async (userId) => {
  const url = `http://localhost:3000/api/users/${userId}/favorites`;
  try{
  const response = await fetch(url);
  const favorites = await response.json();
  return favorites;
  } catch(err) {
    const error = "Failed to get favorites";
    throw error;
  }
}

const removeFavorite = async (favorite) => {
  debugger
  const url = `http://localhost:3000/api/users/${favorite.user_id}/favorites/${favorite.movie_id}`
  try{
    const response = await fetch(url, {
      method: 'DELETE',
      headers: {
        "Content-Type": "application/json",
        },
      body: JSON.stringify(favorite)

    })
    const data = await response.json();
  } catch(err) {
    const error = "Failed to remove favorite";
    throw error;
  }
}


export {
  fetchRecentFilms, 
  addUserFetch, 
  fetchUser,
  postFavoriteToDb,
  getUserFavorites,
  removeFavorite
};