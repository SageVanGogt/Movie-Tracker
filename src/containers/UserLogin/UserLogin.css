.user-login-form {
  display: inline-block;
  margin: 20px;
  border-left: 2px solid #FFF;
  padding-left: 40px;
  height: 250px;
}

.form-field {
  display: block;
  width: 350px;
  height: 30px;
  margin: 15px;
  padding-left: 10px;
}

.submit-button {
  display: block;
  width: 350px;
  background-color: #a108ff;
  color: #FFF;
  height: 30px;
  margin: 15px;
  box-sizing: content-box;
  border: none;
  cursor: pointer;
}

input[type="submit"]:hover {
  background-color: #b748fc;
}