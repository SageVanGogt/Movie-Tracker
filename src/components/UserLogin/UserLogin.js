import React, { Component } from "react";
import { fetchUsers } from "../../apiCall/apiCall";

class UserLogin extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: "",
      password: ""
    };
  }

  handleChange = (event) => {
    const {name, value} = event.target
    this.setState({
        [name]: value
    })
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    console.log(this.state)
    try {
      const response = await fetchUsers(this.state)
      console.log(response)
      this.setState({
        email: "",
        password: ""
      })
    } catch(err) {
      const error = "Failed to grab user data";
      throw error
    }
  };

  render() {
    return (
      <form type="submit" onSubmit={this.handleSubmit}>
        <input
          type="text"
          value={this.state.email}
          name="email"
          onChange={this.handleChange}
        />
        <input
          type="text"
          value={this.state.password}
          name="password"
          onChange={this.handleChange}
        />
         < input
         type = "submit"
         />
      </form>
    );
  }
}

export default UserLogin