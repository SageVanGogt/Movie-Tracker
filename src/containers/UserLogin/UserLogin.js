import React, { Component } from "react";
import { fetchUser } from "../../apiCall/apiCall";
import { updateStoreUser } from './../../actions/index'
import { connect } from 'react-redux'

export class UserLogin extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: "",
      password: ""
    };
  }

  handleChange = (event) => {
    const {name, value} = event.target
    this.setState({
        [name]: value
    })
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    try {
      const response = await fetchUser(this.state)
      this.props.handleLogin(response.data)
      this.setState({
        email: "",
        password: ""
      })
    } catch(err) {
      const error = "Failed to grab user data";
      throw error
    }
  };

  render() {
    return (
      <form type="submit" onSubmit={this.handleSubmit}>
        <input
          type="text"
          value={this.state.email}
          name="email"
          onChange={this.handleChange}
          placeholder="email"
        />
        <input
          type="password"
          value={this.state.password}
          name="password"
          onChange={this.handleChange}
          placeholder="password"          
        />
         < input
         type = "submit"
         />
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  handleLogin: (user) => dispatch(updateStoreUser(user))
})

export default connect(null, mapDispatchToProps)(UserLogin);