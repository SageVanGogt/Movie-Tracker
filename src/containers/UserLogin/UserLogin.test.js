import {UserLogin, mapDispatchToProps} from "./UserLogin";
import React from "react";
import { shallow } from "enzyme";
import { fetchUser, getUserFavorites } from "./../../apiCall/apiCall";


jest.mock('./../../apiCall/apiCall');

describe("UserLogin", () => {
  let wrapper;
  let mockHandleLogin;
  let mockPopulateFavorites;
  beforeEach(() => {
    mockPopulateFavorites = jest.fn();
    mockHandleLogin= jest.fn();
    wrapper = shallow(<UserLogin 
      handleLogin={mockHandleLogin} 
      populateFavorites={mockPopulateFavorites}
      />);
  });

  it("has a default state", () => {
    let expected = {
      email: "",
      password: ""
    };

    expect(wrapper.state()).toEqual(expected);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should update state on change", () => {
    let mockEvent = {
      target: { name: "email", value: "great" }
    };
    let expected = {
      email: "great",
      password: ""
    };

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state()).toEqual(expected);
  });

  it('should calls fetchUser callback after adding user', async () => {
    let mockEvent = {preventDefault: jest.fn()}
    Promise.resolve(wrapper.instance().handleSubmit(mockEvent))
    
    expect(fetchUser).toHaveBeenCalledWith(wrapper.state());
  });

  describe('mapDispatchToProps', () => {
    let wrapper;
    let mockHandleSignIn;
    let mockPopulateFavorites;
    beforeEach(() => {
      mockPopulateFavorites = jest.fn();
      mockHandleSignIn= jest.fn();
      wrapper = shallow(<UserLogin 
        handleSignIn={mockHandleSignIn} 
        populateFavorites={mockPopulateFavorites}
      />);
    });

    it('should call dispatch on handleLogin with the correct params', () => {
      const mockDispatch = jest.fn()
      const mappedProps = mapDispatchToProps(mockDispatch)
      const mockUser = {
       id: 1,
       name: "Doc"
      }
      const mockAction = {
        type: 'ADD_USER',
        user_id: 1,
        name: "Doc"
      }

      mappedProps.handleLogin(mockUser)

      expect(mockDispatch).toHaveBeenCalledWith(mockAction)
    })

    it('should call dispatch on populateFavorites with the correct params', () => {
      const mockDispatch = jest.fn()
      const mappedProps = mapDispatchToProps(mockDispatch)
      const mockAction = {
        type: "ADD_FAVORITES",
        favorites: [{}, {}]
      }
      const mockFavorites = [{},{}]
  
      mappedProps.populateFavorites(mockFavorites)
  
      expect(mockDispatch).toHaveBeenCalledWith(mockAction)
    });

     it('should call dispatch on handleError with the correct params', () => {
       const mockDispatch = jest.fn()
       const mappedProps = mapDispatchToProps(mockDispatch)
       const mockAction = {
         type: "ADD_ERROR",
         message: "Password does not match email"
       }
       const mockError = "Password does not match email";

       mappedProps.handleError(mockError)

       expect(mockDispatch).toHaveBeenCalledWith(mockAction)
     });
  });


});


