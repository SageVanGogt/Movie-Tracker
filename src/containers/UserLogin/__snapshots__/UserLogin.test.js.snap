// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserLogin should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserLogin
    handleLogin={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="text"
          value=""
/>,
        <input
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="password"
          value=""
/>,
        <input
          type="submit"
/>,
      ],
      "onSubmit": [Function],
      "type": "submit",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "email",
          "onChange": [Function],
          "placeholder": "email",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "password",
          "onChange": [Function],
          "placeholder": "password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "type": "submit",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="text"
            value=""
/>,
          <input
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
/>,
          <input
            type="submit"
/>,
        ],
        "onSubmit": [Function],
        "type": "submit",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "email",
            "onChange": [Function],
            "placeholder": "email",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
